File: swagger-codegen/java/src/main/java/io/swagger/client/model/DispatchRoute.java
/*
 * Samsara API
 * # Introduction  Samsara provides API endpoints for interacting with Samsara Cloud, so that you can build powerful applications and custom solutions with sensor data. Samsara has endpoints available to track and analyze sensors, vehicles, and entire fleets.  The Samsara Cloud API is a [RESTful API](https://en.wikipedia.org/wiki/Representational_state_transfer) accessed by an [HTTP](https://en.wikipedia.org/wiki/Hypertext_Transfer_Protocol) client such as wget or curl, or HTTP libraries of most modern programming languages including python, ruby, java. We use built-in HTTP features, like HTTP authentication and HTTP verbs, which are understood by off-the-shelf HTTP clients. We allow you to interact securely with our API from a client-side web application (though you should never expose your secret API key). [JSON](http://www.json.org/) is returned by all API responses, including errors. If you’re familiar with what you can build with a REST API, the following API reference guide will be your go-to resource.  API access to the Samsara cloud is available to all Samsara administrators. To start developing with Samsara APIs you will need to [obtain your API keys](#section/Authentication) to authenticate your API requests.  If you have any questions you can reach out to us on [support@samsara.com](mailto:support@samsara.com)  # Endpoints  All our APIs can be accessed through HTTP requests to URLs like:  ```curl https://api.samsara.com/<version>/<endpoint> ```  All our APIs are [versioned](#section/Versioning). If we intend to make breaking changes to an API which either changes the response format or request parameter, we will increment the version.  # Authentication  To authenticate your API request you will need to include your secret token. You can manage your API tokens in the [Dashboard](https://cloud.samsara.com). They are visible under `Settings->Organization->API Tokens`.  Your API tokens carry many privileges, so be sure to keep them secure. Do not share your secret API tokens in publicly accessible areas such as GitHub, client-side code, and so on.  Authentication to the API is performed via [HTTP Basic Auth](https://en.wikipedia.org/wiki/Basic_access_authentication). Provide your API token as the basic access_token value in the URL. You do not need to provide a password.  ```curl https://api.samsara.com/<version>/<endpoint>?access_token={access_token} ```  All API requests must be made over [HTTPS](https://en.wikipedia.org/wiki/HTTPS). Calls made over plain HTTP or without authentication will fail.  # Request Methods  Our API endpoints use [HTTP request methods](https://en.wikipedia.org/wiki/Hypertext_Transfer_Protocol#Request_methods) to specify the desired operation to be performed. The documentation below specified request method supported by each endpoint and the resulting action.  ## GET  GET requests are typically used for fetching data (like data for a particular driver).  ## POST  POST requests are typically used for creating or updating a record (like adding new tags to the system). With that being said, a few of our POST requests can be used for fetching data (like current location data of your fleet).  ## PUT  PUT requests are typically used for updating an existing record (like updating all devices associated with a particular tag).  ## PATCH  PATCH requests are typically used for modifying an existing record (like modifying a few devices associated with a particular tag).  ## DELETE  DELETE requests are used for deleting a record (like deleting a tag from the system).  # Response Codes  All API requests will respond with appropriate [HTTP status code](https://en.wikipedia.org/wiki/List_of_HTTP_status_codes). Your API client should handle each response class differently.  ## 2XX  These are successful responses and indicate that the API request returned the expected response.  ## 4XX  These indicate that there was a problem with the request like a missing parameter or invalid values. Check the response for specific [error details](#section/Error-Responses). Requests that respond with a 4XX status code, should be modified before retrying.  ## 5XX  These indicate server errors when the server is unreachable or is misconfigured. In this case, you should retry the API request after some delay.  # Error Responses  In case of a 4XX status code, the body of the response will contain information to briefly explain the error reported. To help debugging the error, you can refer to the following table for understanding the error message.  | Status Code | Message | Description | |-------------|----------------|-------------------------------------------------------------------| | 401 | Invalid token | The API token is invalid and could not be authenticated. Please refer to the [authentication section](#section/Authentication). | | 404 | Page not found | The API endpoint being accessed is invalid. | | 400 | Bad request | Default response for an invalid request. Please check the request to make sure it follows the format specified in the documentation. |  # Versioning  All our APIs are versioned. Our current API version is `v1` and we are continuously working on improving it further and provide additional endpoints. If we intend to make breaking changes to an API which either changes the response format or request parameter, we will increment the version. Thus, you can use our current API version worry free.  # FAQs  Check out our [responses to FAQs here](https://kb.samsara.com/hc/en-us/sections/360000538054-APIs). Don’t see an answer to your question? Reach out to us on [support@samsara.com](mailto:support@samsara.com).
 *
 * OpenAPI spec version: 1.0.0
 * 
 *
 * NOTE: This class is auto generated by the swagger code generator program.
 * https://github.com/swagger-api/swagger-codegen.git
 * Do not edit the class manually.
 */


package io.swagger.client.model;

import java.util.Objects;
import java.util.Arrays;
import com.google.gson.TypeAdapter;
import com.google.gson.annotations.JsonAdapter;
import com.google.gson.annotations.SerializedName;
import com.google.gson.stream.JsonReader;
import com.google.gson.stream.JsonWriter;
import io.swagger.annotations.ApiModel;
import io.swagger.annotations.ApiModelProperty;
import io.swagger.client.model.DispatchJob;
import io.swagger.client.model.DispatchRouteBase;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;

/**
 * DispatchRoute
 */
@javax.annotation.Generated(value = "io.swagger.codegen.languages.JavaClientCodegen", date = "2019-03-01T04:19:45.344Z")
public class DispatchRoute {
  @SerializedName("actual_end_ms")
  private Long actualEndMs = null;

  @SerializedName("actual_start_ms")
  private Long actualStartMs = null;

  @SerializedName("driver_id")
  private Long driverId = null;

  @SerializedName("group_id")
  private Long groupId = null;

  @SerializedName("name")
  private String name = null;

  @SerializedName("scheduled_end_ms")
  private Long scheduledEndMs = null;

  @SerializedName("scheduled_meters")
  private Long scheduledMeters = null;

  @SerializedName("scheduled_start_ms")
  private Long scheduledStartMs = null;

  @SerializedName("start_location_address")
  private String startLocationAddress = null;

  @SerializedName("start_location_address_id")
  private Long startLocationAddressId = null;

  @SerializedName("start_location_lat")
  private Double startLocationLat = null;

  @SerializedName("start_location_lng")
  private Double startLocationLng = null;

  @SerializedName("start_location_name")
  private String startLocationName = null;

  @SerializedName("trailer_id")
  private Long trailerId = null;

  @SerializedName("vehicle_id")
  private Long vehicleId = null;

  @SerializedName("dispatch_jobs")
  private List<DispatchJob> dispatchJobs = new ArrayList<DispatchJob>();

  @SerializedName("id")
  private Long id = null;

  public DispatchRoute actualEndMs(Long actualEndMs) {
    this.actualEndMs = actualEndMs;
    return this;
  }

   /**
   * The time in Unix epoch milliseconds that the route actually ended.
   * @return actualEndMs
  **/
  @ApiModelProperty(example = "1462882101000", value = "The time in Unix epoch milliseconds that the route actually ended.")
  public Long getActualEndMs() {
    return actualEndMs;
  }

  public void setActualEndMs(Long actualEndMs) {
    this.actualEndMs = actualEndMs;
  }

  public DispatchRoute actualStartMs(Long actualStartMs) {
    this.actualStartMs = actualStartMs;
    return this;
  }

   /**
   * The time in Unix epoch milliseconds that the route actually started.
   * @return actualStartMs
  **/
  @ApiModelProperty(example = "1462882098000", value = "The time in Unix epoch milliseconds that the route actually started.")
  public Long getActualStartMs() {
    return actualStartMs;
  }

  public void setActualStartMs(Long actualStartMs) {
    this.actualStartMs = actualStartMs;
  }

  public DispatchRoute driverId(Long driverId) {
    this.driverId = driverId;
    return this;
  }

   /**
   * ID of the driver assigned to the dispatch route. Note that driver_id and vehicle_id are mutually exclusive. If neither is specified, then the route is unassigned.
   * @return driverId
  **/
  @ApiModelProperty(example = "555", value = "ID of the driver assigned to the dispatch route. Note that driver_id and vehicle_id are mutually exclusive. If neither is specified, then the route is unassigned.")
  public Long getDriverId() {
    return driverId;
  }

  public void setDriverId(Long driverId) {
    this.driverId = driverId;
  }

  public DispatchRoute groupId(Long groupId) {
    this.groupId = groupId;
    return this;
  }

   /**
   * ID of the group if the organization has multiple groups (optional).
   * @return groupId
  **/
  @ApiModelProperty(example = "101", value = "ID of the group if the organization has multiple groups (optional).")
  public Long getGroupId() {
    return groupId;
  }

  public void setGroupId(Long groupId) {
    this.groupId = groupId;
  }

  public DispatchRoute name(String name) {
    this.name = name;
    return this;
  }

   /**
   * Descriptive name of this route.
   * @return name
  **/
  @ApiModelProperty(example = "Bid #123", required = true, value = "Descriptive name of this route.")
  public String getName() {
    return name;
  }

  public void setName(String name) {
    this.name = name;
  }

  public DispatchRoute scheduledEndMs(Long scheduledEndMs) {
    this.scheduledEndMs = scheduledEndMs;
    return this;
  }

   /**
   * The time in Unix epoch milliseconds that the last job in the route is scheduled to end.
   * @return scheduledEndMs
  **/
  @ApiModelProperty(example = "1462881998034", required = true, value = "The time in Unix epoch milliseconds that the last job in the route is scheduled to end.")
  public Long getScheduledEndMs() {
    return scheduledEndMs;
  }

  public void setScheduledEndMs(Long scheduledEndMs) {
    this.scheduledEndMs = scheduledEndMs;
  }

  public DispatchRoute scheduledMeters(Long scheduledMeters) {
    this.scheduledMeters = scheduledMeters;
    return this;
  }

   /**
   * The distance expected to be traveled for this route in meters.
   * @return scheduledMeters
  **/
  @ApiModelProperty(example = "10000", value = "The distance expected to be traveled for this route in meters.")
  public Long getScheduledMeters() {
    return scheduledMeters;
  }

  public void setScheduledMeters(Long scheduledMeters) {
    this.scheduledMeters = scheduledMeters;
  }

  public DispatchRoute scheduledStartMs(Long scheduledStartMs) {
    this.scheduledStartMs = scheduledStartMs;
    return this;
  }

   /**
   * The time in Unix epoch milliseconds that the route is scheduled to start.
   * @return scheduledStartMs
  **/
  @ApiModelProperty(example = "1462881998034", required = true, value = "The time in Unix epoch milliseconds that the route is scheduled to start.")
  public Long getScheduledStartMs() {
    return scheduledStartMs;
  }

  public void setScheduledStartMs(Long scheduledStartMs) {
    this.scheduledStartMs = scheduledStartMs;
  }

  public DispatchRoute startLocationAddress(String startLocationAddress) {
    this.startLocationAddress = startLocationAddress;
    return this;
  }

   /**
   * The address of the route&#39;s starting location, as it would be recognized if provided to maps.google.com. Optional if a valid start location address ID is provided.
   * @return startLocationAddress
  **/
  @ApiModelProperty(example = "123 Main St, Philadelphia, PA 19106", value = "The address of the route's starting location, as it would be recognized if provided to maps.google.com. Optional if a valid start location address ID is provided.")
  public String getStartLocationAddress() {
    return startLocationAddress;
  }

  public void setStartLocationAddress(String startLocationAddress) {
    this.startLocationAddress = startLocationAddress;
  }

  public DispatchRoute startLocationAddressId(Long startLocationAddressId) {
    this.startLocationAddressId = startLocationAddressId;
    return this;
  }

   /**
   * ID of the start location associated with an address book entry. Optional if valid values are provided for start location address or latitude/longitude. If a valid start location address ID is provided, address/latitude/longitude will be used from the address book entry. Name of the address book entry will only be used if the start location name is not provided.
   * @return startLocationAddressId
  **/
  @ApiModelProperty(example = "67890", value = "ID of the start location associated with an address book entry. Optional if valid values are provided for start location address or latitude/longitude. If a valid start location address ID is provided, address/latitude/longitude will be used from the address book entry. Name of the address book entry will only be used if the start location name is not provided.")
  public Long getStartLocationAddressId() {
    return startLocationAddressId;
  }

  public void setStartLocationAddressId(Long startLocationAddressId) {
    this.startLocationAddressId = startLocationAddressId;
  }

  public DispatchRoute startLocationLat(Double startLocationLat) {
    this.startLocationLat = startLocationLat;
    return this;
  }

   /**
   * Latitude of the start location in decimal degrees. Optional if a valid start location address ID is provided.
   * @return startLocationLat
  **/
  @ApiModelProperty(example = "123.456", value = "Latitude of the start location in decimal degrees. Optional if a valid start location address ID is provided.")
  public Double getStartLocationLat() {
    return startLocationLat;
  }

  public void setStartLocationLat(Double startLocationLat) {
    this.startLocationLat = startLocationLat;
  }

  public DispatchRoute startLocationLng(Double startLocationLng) {
    this.startLocationLng = startLocationLng;
    return this;
  }

   /**
   * Longitude of the start location in decimal degrees. Optional if a valid start location address ID is provided.
   * @return startLocationLng
  **/
  @ApiModelProperty(example = "37.459", value = "Longitude of the start location in decimal degrees. Optional if a valid start location address ID is provided.")
  public Double getStartLocationLng() {
    return startLocationLng;
  }

  public void setStartLocationLng(Double startLocationLng) {
    this.startLocationLng = startLocationLng;
  }

  public DispatchRoute startLocationName(String startLocationName) {
    this.startLocationName = startLocationName;
    return this;
  }

   /**
   * The name of the route&#39;s starting location. If provided, it will take precedence over the name of the address book entry.
   * @return startLocationName
  **/
  @ApiModelProperty(example = "ACME Inc. Philadelphia HQ", value = "The name of the route's starting location. If provided, it will take precedence over the name of the address book entry.")
  public String getStartLocationName() {
    return startLocationName;
  }

  public void setStartLocationName(String startLocationName) {
    this.startLocationName = startLocationName;
  }

  public DispatchRoute trailerId(Long trailerId) {
    this.trailerId = trailerId;
    return this;
  }

   /**
   * ID of the trailer assigned to the dispatch route. Note that trailers can only be assigned to routes that have a Vehicle or Driver assigned to them.
   * @return trailerId
  **/
  @ApiModelProperty(example = "666", value = "ID of the trailer assigned to the dispatch route. Note that trailers can only be assigned to routes that have a Vehicle or Driver assigned to them.")
  public Long getTrailerId() {
    return trailerId;
  }

  public void setTrailerId(Long trailerId) {
    this.trailerId = trailerId;
  }

  public DispatchRoute vehicleId(Long vehicleId) {
    this.vehicleId = vehicleId;
    return this;
  }

   /**
   * ID of the vehicle assigned to the dispatch route. Note that vehicle_id and driver_id are mutually exclusive. If neither is specified, then the route is unassigned.
   * @return vehicleId
  **/
  @ApiModelProperty(example = "444", value = "ID of the vehicle assigned to the dispatch route. Note that vehicle_id and driver_id are mutually exclusive. If neither is specified, then the route is unassigned.")
  public Long getVehicleId() {
    return vehicleId;
  }

  public void setVehicleId(Long vehicleId) {
    this.vehicleId = vehicleId;
  }

  public DispatchRoute dispatchJobs(List<DispatchJob> dispatchJobs) {
    this.dispatchJobs = dispatchJobs;
    return this;
  }

  public DispatchRoute addDispatchJobsItem(DispatchJob dispatchJobsItem) {
    this.dispatchJobs.add(dispatchJobsItem);
    return this;
  }

   /**
   * The dispatch jobs associated with this route.
   * @return dispatchJobs
  **/
  @ApiModelProperty(required = true, value = "The dispatch jobs associated with this route.")
  public List<DispatchJob> getDispatchJobs() {
    return dispatchJobs;
  }

  public void setDispatchJobs(List<DispatchJob> dispatchJobs) {
    this.dispatchJobs = dispatchJobs;
  }

  public DispatchRoute id(Long id) {
    this.id = id;
    return this;
  }

   /**
   * ID of the Samsara dispatch route.
   * @return id
  **/
  @ApiModelProperty(example = "556", required = true, value = "ID of the Samsara dispatch route.")
  public Long getId() {
    return id;
  }

  public void setId(Long id) {
    this.id = id;
  }


  @Override
  public boolean equals(java.lang.Object o) {
    if (this == o) {
      return true;
    }
    if (o == null || getClass() != o.getClass()) {
      return false;
    }
    DispatchRoute dispatchRoute = (DispatchRoute) o;
    return Objects.equals(this.actualEndMs, dispatchRoute.actualEndMs) &&
        Objects.equals(this.actualStartMs, dispatchRoute.actualStartMs) &&
        Objects.equals(this.driverId, dispatchRoute.driverId) &&
        Objects.equals(this.groupId, dispatchRoute.groupId) &&
        Objects.equals(this.name, dispatchRoute.name) &&
        Objects.equals(this.scheduledEndMs, dispatchRoute.scheduledEndMs) &&
        Objects.equals(this.scheduledMeters, dispatchRoute.scheduledMeters) &&
        Objects.equals(this.scheduledStartMs, dispatchRoute.scheduledStartMs) &&
        Objects.equals(this.startLocationAddress, dispatchRoute.startLocationAddress) &&
        Objects.equals(this.startLocationAddressId, dispatchRoute.startLocationAddressId) &&
        Objects.equals(this.startLocationLat, dispatchRoute.startLocationLat) &&
        Objects.equals(this.startLocationLng, dispatchRoute.startLocationLng) &&
        Objects.equals(this.startLocationName, dispatchRoute.startLocationName) &&
        Objects.equals(this.trailerId, dispatchRoute.trailerId) &&
        Objects.equals(this.vehicleId, dispatchRoute.vehicleId) &&
        Objects.equals(this.dispatchJobs, dispatchRoute.dispatchJobs) &&
        Objects.equals(this.id, dispatchRoute.id);
  }

  @Override
  public int hashCode() {
    return Objects.hash(actualEndMs, actualStartMs, driverId, groupId, name, scheduledEndMs, scheduledMeters, scheduledStartMs, startLocationAddress, startLocationAddressId, startLocationLat, startLocationLng, startLocationName, trailerId, vehicleId, dispatchJobs, id);
  }


  @Override
  public String toString() {
    StringBuilder sb = new StringBuilder();
    sb.append("class DispatchRoute {\n");
    
    sb.append("    actualEndMs: ").append(toIndentedString(actualEndMs)).append("\n");
    sb.append("    actualStartMs: ").append(toIndentedString(actualStartMs)).append("\n");
    sb.append("    driverId: ").append(toIndentedString(driverId)).append("\n");
    sb.append("    groupId: ").append(toIndentedString(groupId)).append("\n");
    sb.append("    name: ").append(toIndentedString(name)).append("\n");
    sb.append("    scheduledEndMs: ").append(toIndentedString(scheduledEndMs)).append("\n");
    sb.append("    scheduledMeters: ").append(toIndentedString(scheduledMeters)).append("\n");
    sb.append("    scheduledStartMs: ").append(toIndentedString(scheduledStartMs)).append("\n");
    sb.append("    startLocationAddress: ").append(toIndentedString(startLocationAddress)).append("\n");
    sb.append("    startLocationAddressId: ").append(toIndentedString(startLocationAddressId)).append("\n");
    sb.append("    startLocationLat: ").append(toIndentedString(startLocationLat)).append("\n");
    sb.append("    startLocationLng: ").append(toIndentedString(startLocationLng)).append("\n");
    sb.append("    startLocationName: ").append(toIndentedString(startLocationName)).append("\n");
    sb.append("    trailerId: ").append(toIndentedString(trailerId)).append("\n");
    sb.append("    vehicleId: ").append(toIndentedString(vehicleId)).append("\n");
    sb.append("    dispatchJobs: ").append(toIndentedString(dispatchJobs)).append("\n");
    sb.append("    id: ").append(toIndentedString(id)).append("\n");
    sb.append("}");
    return sb.toString();
  }

  /**
   * Convert the given object to string with each line indented by 4 spaces
   * (except the first line).
   */
  private String toIndentedString(java.lang.Object o) {
    if (o == null) {
      return "null";
    }
    return o.toString().replace("\n", "\n    ");
  }

}

File: openapi-generator/csharp-nancyfx/src/Org.OpenAPITools/Models/DispatchRoute.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Text;
using Sharpility.Extensions;
using NodaTime;

namespace Org.OpenAPITools.v1.Models
{
    /// <summary>
    /// DispatchRoute
    /// </summary>
    public sealed class DispatchRoute:  IEquatable<DispatchRoute>
    { 
        /// <summary>
        /// The time in Unix epoch milliseconds that the route actually ended.
        /// </summary>
        public long? ActualEndMs { get; private set; }

        /// <summary>
        /// The time in Unix epoch milliseconds that the route actually started.
        /// </summary>
        public long? ActualStartMs { get; private set; }

        /// <summary>
        /// ID of the driver assigned to the dispatch route. Note that driver_id and vehicle_id are mutually exclusive. If neither is specified, then the route is unassigned.
        /// </summary>
        public long? DriverId { get; private set; }

        /// <summary>
        /// ID of the group if the organization has multiple groups (optional).
        /// </summary>
        public long? GroupId { get; private set; }

        /// <summary>
        /// Descriptive name of this route.
        /// </summary>
        public string Name { get; private set; }

        /// <summary>
        /// The time in Unix epoch milliseconds that the last job in the route is scheduled to end.
        /// </summary>
        public long? ScheduledEndMs { get; private set; }

        /// <summary>
        /// The distance expected to be traveled for this route in meters.
        /// </summary>
        public long? ScheduledMeters { get; private set; }

        /// <summary>
        /// The time in Unix epoch milliseconds that the route is scheduled to start.
        /// </summary>
        public long? ScheduledStartMs { get; private set; }

        /// <summary>
        /// The address of the route&#39;s starting location, as it would be recognized if provided to maps.google.com. Optional if a valid start location address ID is provided.
        /// </summary>
        public string StartLocationAddress { get; private set; }

        /// <summary>
        /// ID of the start location associated with an address book entry. Optional if valid values are provided for start location address or latitude/longitude. If a valid start location address ID is provided, address/latitude/longitude will be used from the address book entry. Name of the address book entry will only be used if the start location name is not provided.
        /// </summary>
        public long? StartLocationAddressId { get; private set; }

        /// <summary>
        /// Latitude of the start location in decimal degrees. Optional if a valid start location address ID is provided.
        /// </summary>
        public double? StartLocationLat { get; private set; }

        /// <summary>
        /// Longitude of the start location in decimal degrees. Optional if a valid start location address ID is provided.
        /// </summary>
        public double? StartLocationLng { get; private set; }

        /// <summary>
        /// The name of the route&#39;s starting location. If provided, it will take precedence over the name of the address book entry.
        /// </summary>
        public string StartLocationName { get; private set; }

        /// <summary>
        /// ID of the trailer assigned to the dispatch route. Note that trailers can only be assigned to routes that have a Vehicle or Driver assigned to them.
        /// </summary>
        public long? TrailerId { get; private set; }

        /// <summary>
        /// ID of the vehicle assigned to the dispatch route. Note that vehicle_id and driver_id are mutually exclusive. If neither is specified, then the route is unassigned.
        /// </summary>
        public long? VehicleId { get; private set; }

        /// <summary>
        /// The dispatch jobs associated with this route.
        /// </summary>
        public List<DispatchJob> DispatchJobs { get; private set; }

        /// <summary>
        /// ID of the Samsara dispatch route.
        /// </summary>
        public long? Id { get; private set; }


        /// <summary>
        /// Empty constructor required by some serializers.
        /// Use DispatchRoute.Builder() for instance creation instead.
        /// </summary>
        [Obsolete]
        public DispatchRoute()
        {
        }

        private DispatchRoute(long? ActualEndMs, long? ActualStartMs, long? DriverId, long? GroupId, string Name, long? ScheduledEndMs, long? ScheduledMeters, long? ScheduledStartMs, string StartLocationAddress, long? StartLocationAddressId, double? StartLocationLat, double? StartLocationLng, string StartLocationName, long? TrailerId, long? VehicleId, List<DispatchJob> DispatchJobs, long? Id)
        {
            
            this.ActualEndMs = ActualEndMs;
            
            this.ActualStartMs = ActualStartMs;
            
            this.DriverId = DriverId;
            
            this.GroupId = GroupId;
            
            this.Name = Name;
            
            this.ScheduledEndMs = ScheduledEndMs;
            
            this.ScheduledMeters = ScheduledMeters;
            
            this.ScheduledStartMs = ScheduledStartMs;
            
            this.StartLocationAddress = StartLocationAddress;
            
            this.StartLocationAddressId = StartLocationAddressId;
            
            this.StartLocationLat = StartLocationLat;
            
            this.StartLocationLng = StartLocationLng;
            
            this.StartLocationName = StartLocationName;
            
            this.TrailerId = TrailerId;
            
            this.VehicleId = VehicleId;
            
            this.DispatchJobs = DispatchJobs;
            
            this.Id = Id;
            
        }

        /// <summary>
        /// Returns builder of DispatchRoute.
        /// </summary>
        /// <returns>DispatchRouteBuilder</returns>
        public static DispatchRouteBuilder Builder()
        {
            return new DispatchRouteBuilder();
        }

        /// <summary>
        /// Returns DispatchRouteBuilder with properties set.
        /// Use it to change properties.
        /// </summary>
        /// <returns>DispatchRouteBuilder</returns>
        public DispatchRouteBuilder With()
        {
            return Builder()
                .ActualEndMs(ActualEndMs)
                .ActualStartMs(ActualStartMs)
                .DriverId(DriverId)
                .GroupId(GroupId)
                .Name(Name)
                .ScheduledEndMs(ScheduledEndMs)
                .ScheduledMeters(ScheduledMeters)
                .ScheduledStartMs(ScheduledStartMs)
                .StartLocationAddress(StartLocationAddress)
                .StartLocationAddressId(StartLocationAddressId)
                .StartLocationLat(StartLocationLat)
                .StartLocationLng(StartLocationLng)
                .StartLocationName(StartLocationName)
                .TrailerId(TrailerId)
                .VehicleId(VehicleId)
                .DispatchJobs(DispatchJobs)
                .Id(Id);
        }

        public override string ToString()
        {
            return this.PropertiesToString();
        }

        public override bool Equals(object obj)
        {
            return this.EqualsByProperties(obj);
        }

        public bool Equals(DispatchRoute other)
        {
            return Equals((object) other);
        }

        public override int GetHashCode()
        {
            return this.PropertiesHash();
        }

        /// <summary>
        /// Implementation of == operator for (DispatchRoute.
        /// </summary>
        /// <param name="left">Compared (DispatchRoute</param>
        /// <param name="right">Compared (DispatchRoute</param>
        /// <returns>true if compared items are equals, false otherwise</returns>
        public static bool operator == (DispatchRoute left, DispatchRoute right)
        {
            return Equals(left, right);
        }

        /// <summary>
        /// Implementation of != operator for (DispatchRoute.
        /// </summary>
        /// <param name="left">Compared (DispatchRoute</param>
        /// <param name="right">Compared (DispatchRoute</param>
        /// <returns>true if compared items are not equals, false otherwise</returns>
        public static bool operator != (DispatchRoute left, DispatchRoute right)
        {
            return !Equals(left, right);
        }

        /// <summary>
        /// Builder of DispatchRoute.
        /// </summary>
        public sealed class DispatchRouteBuilder
        {
            private long? _ActualEndMs;
            private long? _ActualStartMs;
            private long? _DriverId;
            private long? _GroupId;
            private string _Name;
            private long? _ScheduledEndMs;
            private long? _ScheduledMeters;
            private long? _ScheduledStartMs;
            private string _StartLocationAddress;
            private long? _StartLocationAddressId;
            private double? _StartLocationLat;
            private double? _StartLocationLng;
            private string _StartLocationName;
            private long? _TrailerId;
            private long? _VehicleId;
            private List<DispatchJob> _DispatchJobs;
            private long? _Id;

            internal DispatchRouteBuilder()
            {
                SetupDefaults();
            }

            private void SetupDefaults()
            {
            }

            /// <summary>
            /// Sets value for DispatchRoute.ActualEndMs property.
            /// </summary>
            /// <param name="value">The time in Unix epoch milliseconds that the route actually ended.</param>
            public DispatchRouteBuilder ActualEndMs(long? value)
            {
                _ActualEndMs = value;
                return this;
            }

            /// <summary>
            /// Sets value for DispatchRoute.ActualStartMs property.
            /// </summary>
            /// <param name="value">The time in Unix epoch milliseconds that the route actually started.</param>
            public DispatchRouteBuilder ActualStartMs(long? value)
            {
                _ActualStartMs = value;
                return this;
            }

            /// <summary>
            /// Sets value for DispatchRoute.DriverId property.
            /// </summary>
            /// <param name="value">ID of the driver assigned to the dispatch route. Note that driver_id and vehicle_id are mutually exclusive. If neither is specified, then the route is unassigned.</param>
            public DispatchRouteBuilder DriverId(long? value)
            {
                _DriverId = value;
                return this;
            }

            /// <summary>
            /// Sets value for DispatchRoute.GroupId property.
            /// </summary>
            /// <param name="value">ID of the group if the organization has multiple groups (optional).</param>
            public DispatchRouteBuilder GroupId(long? value)
            {
                _GroupId = value;
                return this;
            }

            /// <summary>
            /// Sets value for DispatchRoute.Name property.
            /// </summary>
            /// <param name="value">Descriptive name of this route.</param>
            public DispatchRouteBuilder Name(string value)
            {
                _Name = value;
                return this;
            }

            /// <summary>
            /// Sets value for DispatchRoute.ScheduledEndMs property.
            /// </summary>
            /// <param name="value">The time in Unix epoch milliseconds that the last job in the route is scheduled to end.</param>
            public DispatchRouteBuilder ScheduledEndMs(long? value)
            {
                _ScheduledEndMs = value;
                return this;
            }

            /// <summary>
            /// Sets value for DispatchRoute.ScheduledMeters property.
            /// </summary>
            /// <param name="value">The distance expected to be traveled for this route in meters.</param>
            public DispatchRouteBuilder ScheduledMeters(long? value)
            {
                _ScheduledMeters = value;
                return this;
            }

            /// <summary>
            /// Sets value for DispatchRoute.ScheduledStartMs property.
            /// </summary>
            /// <param name="value">The time in Unix epoch milliseconds that the route is scheduled to start.</param>
            public DispatchRouteBuilder ScheduledStartMs(long? value)
            {
                _ScheduledStartMs = value;
                return this;
            }

            /// <summary>
            /// Sets value for DispatchRoute.StartLocationAddress property.
            /// </summary>
            /// <param name="value">The address of the route&#39;s starting location, as it would be recognized if provided to maps.google.com. Optional if a valid start location address ID is provided.</param>
            public DispatchRouteBuilder StartLocationAddress(string value)
            {
                _StartLocationAddress = value;
                return this;
            }

            /// <summary>
            /// Sets value for DispatchRoute.StartLocationAddressId property.
            /// </summary>
            /// <param name="value">ID of the start location associated with an address book entry. Optional if valid values are provided for start location address or latitude/longitude. If a valid start location address ID is provided, address/latitude/longitude will be used from the address book entry. Name of the address book entry will only be used if the start location name is not provided.</param>
            public DispatchRouteBuilder StartLocationAddressId(long? value)
            {
                _StartLocationAddressId = value;
                return this;
            }

            /// <summary>
            /// Sets value for DispatchRoute.StartLocationLat property.
            /// </summary>
            /// <param name="value">Latitude of the start location in decimal degrees. Optional if a valid start location address ID is provided.</param>
            public DispatchRouteBuilder StartLocationLat(double? value)
            {
                _StartLocationLat = value;
                return this;
            }

            /// <summary>
            /// Sets value for DispatchRoute.StartLocationLng property.
            /// </summary>
            /// <param name="value">Longitude of the start location in decimal degrees. Optional if a valid start location address ID is provided.</param>
            public DispatchRouteBuilder StartLocationLng(double? value)
            {
                _StartLocationLng = value;
                return this;
            }

            /// <summary>
            /// Sets value for DispatchRoute.StartLocationName property.
            /// </summary>
            /// <param name="value">The name of the route&#39;s starting location. If provided, it will take precedence over the name of the address book entry.</param>
            public DispatchRouteBuilder StartLocationName(string value)
            {
                _StartLocationName = value;
                return this;
            }

            /// <summary>
            /// Sets value for DispatchRoute.TrailerId property.
            /// </summary>
            /// <param name="value">ID of the trailer assigned to the dispatch route. Note that trailers can only be assigned to routes that have a Vehicle or Driver assigned to them.</param>
            public DispatchRouteBuilder TrailerId(long? value)
            {
                _TrailerId = value;
                return this;
            }

            /// <summary>
            /// Sets value for DispatchRoute.VehicleId property.
            /// </summary>
            /// <param name="value">ID of the vehicle assigned to the dispatch route. Note that vehicle_id and driver_id are mutually exclusive. If neither is specified, then the route is unassigned.</param>
            public DispatchRouteBuilder VehicleId(long? value)
            {
                _VehicleId = value;
                return this;
            }

            /// <summary>
            /// Sets value for DispatchRoute.DispatchJobs property.
            /// </summary>
            /// <param name="value">The dispatch jobs associated with this route.</param>
            public DispatchRouteBuilder DispatchJobs(List<DispatchJob> value)
            {
                _DispatchJobs = value;
                return this;
            }

            /// <summary>
            /// Sets value for DispatchRoute.Id property.
            /// </summary>
            /// <param name="value">ID of the Samsara dispatch route.</param>
            public DispatchRouteBuilder Id(long? value)
            {
                _Id = value;
                return this;
            }


            /// <summary>
            /// Builds instance of DispatchRoute.
            /// </summary>
            /// <returns>DispatchRoute</returns>
            public DispatchRoute Build()
            {
                Validate();
                return new DispatchRoute(
                    ActualEndMs: _ActualEndMs,
                    ActualStartMs: _ActualStartMs,
                    DriverId: _DriverId,
                    GroupId: _GroupId,
                    Name: _Name,
                    ScheduledEndMs: _ScheduledEndMs,
                    ScheduledMeters: _ScheduledMeters,
                    ScheduledStartMs: _ScheduledStartMs,
                    StartLocationAddress: _StartLocationAddress,
                    StartLocationAddressId: _StartLocationAddressId,
                    StartLocationLat: _StartLocationLat,
                    StartLocationLng: _StartLocationLng,
                    StartLocationName: _StartLocationName,
                    TrailerId: _TrailerId,
                    VehicleId: _VehicleId,
                    DispatchJobs: _DispatchJobs,
                    Id: _Id
                );
            }

            private void Validate()
            { 
                if (_Name == null)
                {
                    throw new ArgumentException("Name is a required property for DispatchRoute and cannot be null");
                } 
                if (_ScheduledEndMs == null)
                {
                    throw new ArgumentException("ScheduledEndMs is a required property for DispatchRoute and cannot be null");
                } 
                if (_ScheduledStartMs == null)
                {
                    throw new ArgumentException("ScheduledStartMs is a required property for DispatchRoute and cannot be null");
                } 
                if (_DispatchJobs == null)
                {
                    throw new ArgumentException("DispatchJobs is a required property for DispatchRoute and cannot be null");
                } 
                if (_Id == null)
                {
                    throw new ArgumentException("Id is a required property for DispatchRoute and cannot be null");
                } 
            }
        }

        
    }
}